### ISSUE TITLE
Neural Networks 101 - Artificial Intelligence

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4445]
Neural Networks 101 - Artificial Intelligence<br>
<br>
http://bit.ly/2x5piUe<br>
<br>
April 4, 2018
            Posted By 
mrsaturn01<br>
<br>
<div><section><p> <strong>Intention</strong> </p> <p> This article is intended to give a brief look into the logistics of machine learning. Do not expect to become an expert on the field just by reading this. However, I hope that the article goes into just enough detail so that it sparks your interest in learning more about AI and how it can be applied to various fields such as games. Once you finish reading the article, I recommend looking at the resources posted below. If you have any questions, feel free to message me on Twitter <span>@adityaXharsh.</span> </p> <p> <strong>How Neural Networks Work</strong> </p> <p> Neural networks work by using a system of receiving inputs, sending outputs, and performing self-corrections based on the difference between the output and expected output, also known as the <em>cost</em>. </p> <p> Neural networks are composed of <em>neurons</em>, which in turn compose <em>layers</em>, or collections of neurons. For example, there is an input layer and an output layer. In between the these two layers, there are layers known as <em>hidden layers</em>. These layers allow for more complex and nuanced behavior by the neural network. A neural network can be thought of as a multi-tier cake: the first tier of the cake represents the input, the tiers in between, or lack thereof, represent the hidden layers, and the last tier represents the output. </p> <p> The two mechanisms of learning are <em>Forward Propagation</em> and <em>Backward Propagation</em>. Forward Propagation uses linear algebra for calculating what the activation of each neuron of the next layer should be, and then pushing, or propagating, those values forward. Backward Propagation uses calculus to determine what values in the network need to be changed in order to bring the output closer to the expected output. </p> <p> <strong>Forward Propagation</strong> </p> <p> <img alt="ann_3.gif" src="https://www.gamedev.net/applications/core/interface/imageproxy/imageproxy.php?img=https://picoledelimao.github.io/images/posts/ann_3.gif&amp;key=7331cdad42a9fa5d75cfb463984d060e58c38a1b7f49e4e46cae9aa789c93a77"></p> <p> As can be seen from the gif above, each layer is composed of multiple neurons, and each neuron is connected to every other neuron of the following and previous layer, save for the input and output layers since they are not surrounding by layers from both sides. </p> <p> <img alt="image.png.2b7512473ace465ef5940279ed53e798.png" src="https://www.gamedev.net/uploads/monthly_2018_04/image.png.2b7512473ace465ef5940279ed53e798.png"></p> <p> To put it simply, a neural network represents a collection of activations, weights, and biases. They can be defined as: </p> <ul>
<li> <em>Activat</em><em>ion</em>: A value representing how strongly a neuron is firing. </li> <li> <em>Weight</em>: How strong the connection is between two neurons. Affects how much of the activation is propagated onto the next layer. </li> <li> <em>Bias</em>: A minimum threshold for whether or not the current neuron's activation and weight should affect the next neuron's activation. </li> </ul>
<p> Each neuron has an activation and a bias. Every connection to every neuron is represented as a weight. The activations, weights, biases, and connections can be represented using matrices. Activations are calculated using this formula: </p> <p> <a href="https://www.gamedev.net/uploads/monthly_2018_04/activ.png.66bd8c0370f522d697b474d5c99ee9aa.png"><img alt="activ.thumb.png.68eafd5e469e68652f7ddc3762b3524b.png" src="https://www.gamedev.net/uploads/monthly_2018_04/activ.thumb.png.68eafd5e469e68652f7ddc3762b3524b.png"></a> </p> <p> After the inner portion of the function has been computed, the resulting matrix gets pumped into a special function known as the <em>Sigmoid Function</em>. The sigmoid is defined as: </p> <p> <img alt="sig.png.e58b7637466f1e2b3a6f2dfe5d1c8fd1.png" src="https://www.gamedev.net/uploads/monthly_2018_04/sig.png.e58b7637466f1e2b3a6f2dfe5d1c8fd1.png"></p> <p> The sigmoid function is handy since its output is locked between a range of zero and one. This process is repeated until the activations of the output neurons have been calculated. </p> <p> <strong>Backward Propagation</strong> </p> <p> The process of a neural network performing self-correction is referred to as Backward Propagation or backprop. This article will not go into detail about backprop since it can be a confusing topic. To summarize, the algorithm uses a technique in calculus known as <em>Gradient Descent</em>. Given a plane in an infinite number of dimensions, the direction of change that minimizes the error must be found. The goal of using gradient descent is to modify the weights and biases such that the error in the network approaches zero. </p> <p> <img alt="engineeronadisk-12.gif" src="https://www.gamedev.net/applications/core/interface/imageproxy/imageproxy.php?img=http://engineeronadisk.com/V2/hugh_jack_masters/engineeronadisk-12.gif&amp;key=b11e18e76dc58eb0d756ee9ad7db67631572dbc28bb393090ead730c90992b4f"></p> <p> Furthermore, you can find the cost, or error, of a network using this formula: </p> <p> <a href="https://www.gamedev.net/uploads/monthly_2018_04/image.png.f53bce7319b7a6d84f5d7c924206e0b7.png"><img alt="image.thumb.png.c8a0b7acacd5acdf9c12a6cd03c13e4e.png" src="https://www.gamedev.net/uploads/monthly_2018_04/image.thumb.png.c8a0b7acacd5acdf9c12a6cd03c13e4e.png"></a> </p> <p> Unlike forward propagation, which is done from input to output, backward propagation goes from output to input. For every activation, find the error in that neuron, how much of a role it played in the error of the output, and adjust accordingly. This technique uses concepts such as the chain rule, partial derivatives, and multi-variate calculus; therefore, it's a good idea to brush up on one's calculus skills. </p> <p> <strong>High Level Algorithm</strong> </p> <ol>
<li> Initialize matrices for weights and biases for all layers to a random decimal number between -1 and 1. </li> <li> Propagate input through the network. </li> <li> Compare output with the expected output. </li> <li> Backwards propagate the correction back into the network. </li> <li> Repeat this for N number of training samples. </li> </ol>
<p> <img alt="UnsteadyUnkemptKangaroo-size_restricted." src="https://www.gamedev.net/applications/core/interface/imageproxy/imageproxy.php?img=https://thumbs.gfycat.com/UnsteadyUnkemptKangaroo-size_restricted.gif&amp;key=8cd5130ddf79dbfdbfd82c33f04f13b351ac27877a55239040e657109b481de9"></p> <p> <strong>Source Code</strong> </p> <p> If you're interested in looking into the guts of a neural network, check out AI Chan! It's a simple to integrate library for machine learning I wrote in C++. Feel free to learn from it and use it in your own projects. </p> <p> <a href="https://bitbucket.org/mrsaturnsan/aichan/overview" rel="external nofollow">https://bitbucket.org/mrsaturnsan/aichan/</a> </p> <p> <strong>Resources</strong> </p> <p> <a href="http://neuralnetworksanddeeplearning.com/" rel="external nofollow">http://neuralnetworksanddeeplearning.com/</a> </p> <p> <a href="https://www.youtube.com/channel/UCWN3xxRkmTPmbKwht9FuE5A" rel="external nofollow">https://www.youtube.com/channel/UCWN3xxRkmTPmbKwht9FuE5A</a> </p> <p>   </p> <p>   </p> </section></div><br>
<br>
<br>
<br>
<br>
<br>
via GameDev.net https://www.gamedev.net<br>
<br>
June 20, 2019 at 09:56PM
